src/client.cpp

```diff
@@ -1386,3 +1458,179 @@ int CClient::EstimatedOverallDelay ( const int iPingTimeMs )
 }
+
+// Management of Client Channels and mapping to/from Server Channels
+
+void CClient::ClearClientChannels()
+{
+    QMutexLocker locker ( &MutexChannels );
+
+    iActiveChannels = 0;
+    iJoinSequence   = 0;
+
+    for ( int i = 0; i < MAX_NUM_CHANNELS; i++ )
+    {
+        clientChannels[i].iServerChannelID = INVALID_INDEX;
+        // all other fields will be initialised on channel allocation
+
+        clientChannelIDs[i] = INVALID_INDEX;
+    }
+
+    qInfo() << "> Client channel list cleared";
+}
+
+void CClient::FreeClientChannel ( const int iServerChannelID )
+{
+    QMutexLocker locker ( &MutexChannels );
+
+    if ( iServerChannelID == INVALID_INDEX || iServerChannelID >= MAX_NUM_CHANNELS )
+    {
+        return;
+    }
+
+    const int iClientChannelID = clientChannelIDs[iServerChannelID];
+
+    Q_ASSERT ( clientChannels[iClientChannelID].iServerChannelID == iServerChannelID );
+
+    clientChannelIDs[iServerChannelID]                = INVALID_INDEX;
+    clientChannels[iClientChannelID].iServerChannelID = INVALID_INDEX;
+
+    iActiveChannels -= 1;
+
+    qInfo() << qUtf8Printable ( QString ( "> Freed client ch %1 for server ch %2; chan count = %3" )
+                                    .arg ( iClientChannelID )
+                                    .arg ( iServerChannelID )
+                                    .arg ( iActiveChannels ) );
+}
+
+// find, and optionally create, a client channel for the supplied server channel ID
+// returns a client channel ID or INVALID_INDEX
+int CClient::FindClientChannel ( const int iServerChannelID, const bool bCreateIfNew )
+{
+    QMutexLocker locker ( &MutexChannels );
+
+    if ( iServerChannelID == INVALID_INDEX || iServerChannelID >= MAX_NUM_CHANNELS )
+    {
+        return INVALID_INDEX;
+    }
+
+    int iClientChannelID = clientChannelIDs[iServerChannelID];
+
+    if ( iClientChannelID != INVALID_INDEX )
+    {
+        Q_ASSERT ( clientChannels[iClientChannelID].iServerChannelID == iServerChannelID );
+
+        return iClientChannelID;
+    }
+
+    // no matching client channel - create new one if requested
+    if ( bCreateIfNew )
+    {
+        // search clientChannels[] for a free client channel
+        for ( iClientChannelID = 0; iClientChannelID < MAX_NUM_CHANNELS; iClientChannelID++ )
+        {
+            CClientChannel* clientChan = &clientChannels[iClientChannelID];
+
+            if ( clientChan->iServerChannelID == INVALID_INDEX )
+            {
+                clientChan->iServerChannelID = iServerChannelID;
+                clientChan->iJoinSequence    = ++iJoinSequence;
+
+                clientChan->oldGain = clientChan->newGain = 1.0f;
+
+                clientChan->level = 0;
+
+                clientChannelIDs[iServerChannelID] = iClientChannelID;
+
+                iActiveChannels += 1;
+
+                qInfo() << qUtf8Printable ( QString ( "> Alloc client ch %1 for server ch %2; chan count = %3" )
```

Comment:
Probably don't want this except for debugging - a busy server will cause a client to be issuing log messages rather a lot...

Reply:
Now that I'm satisfied with the code operation, I'm happy to comment these out, or to change them to `qDebug()`. In the latter case, they would be output from an executable that was built with `make debug`, but together with a LOT of noise from the Qt graphics code!